modules/raft/tech-notes/rebalance.md

```diff
@@ -0,0 +1,71 @@
+        else:
+            if zoneId.assignments.pending != newAssignments
+                zoneId.assignments.planned = newAssignments
+                zoneId.assignments.change.revision = configurationUpdate.revision
+            else
+                remove(zoneId.assignments.planned)
+    else:
+        skip
+```
+### 2. Wait for the all needed replicas to start rebalance routine
+It looks like we can reuse the mechanism of AwaitReplicaRequest:
+- PrimaryReplica send an AwaitReplicaRequest to all new replicas.
+- When all answers received, rebalance can be started 
+
+### 3. Replication group rebalance
+Let's zoom to the details of PrimaryReplica and replication group communication for the RAFT case:
+![](images/primaryReplica.svg)
+
+* Any replication member can has in-flight RO transactions. But at the same time, if it is not a member of new topology, it will not receive updates of safe time, so these RO transactions will be failed by timeout. It is not an issue for correctness, but we want to optimise this in future to avoid redundant transaction fails and retries [TODO].
+
+#### 3.1 Notification about the new leader and rebalance events
+Current rebalance algorithm based on the metastore invokes and local rebalance listeners.
+
+But for the new one we have an idea, which doesn't need the metastore at all:
+- On rebalanceDone/rebalanceError/leaderElected events the local event listener send a message to PrimaryReplica with the description of event
+- If PrimaryReplica is not available - we should retry send, until leader didn't find himself outdated (in this case, new leader will send leaderElected event to PrimaryReplica and receives the rebalance request again.
+
+### 4. Stop the redundant replicas and update replicas clients
+Here we can accidentally stop PrimaryReplica, so we need to use the algorithm of a graceful PrimaryReplica transfer, if needed.
+
+Also, we need to update the replication protocol clients (RaftGroupService, for example) on each Replica.
+
+### Failover logic
+The main idea of failover process: every rebalance request PlacementDriver->PrimaryReplica or PrimaryReplica->ReplicationGroup must be idempotent. So, redundant request in the worst case should be just answered by positive answer (just like rebalance is already done).
+
+After that we can prepare the following logic:
+- On every new PD leader elected - it must check the direct value (not the locally cached one) of `zoneId.assignment.pending` keys and send RebalanceRequest to needed PrimaryReplicas and then listen updates from the last revision.
+- On every PrimaryReplica reelection by PD it must send the new RebalanceRequest to PrimaryReplica, if pending key is not empty. 
+- On every leader reelection (for the leader oriented protocols) inside the replication group - leader send leaderElected event to PrimaryReplica, which force PrimaryReplica to send RebalanceRequest to RAFT leader again.
```

Comment:
seems that RAFT shouldn't be mentioned here - as the first part of the sentence is about leader oriented protocols in general